stellargraph/core/tf_utils.py

```diff
@@ -0,0 +1,63 @@
+import tensorflow as tf
+from tensorflow.keras import backend as K
+from ..core.experimental import experimental
+
+
+@experimental(reason="lack of unit tests")
+def partial_powers(one_hot_encoded_row, Aadj_T, num_powers=5):
```

Comment:
The paper talks about doing an approximate SVD to compute this; I assume this code is not doing that?

I suspect that this is going to be O(k * n^3) where n is the number of nodes.

Reply:
Yeah you're write that the code isn't doing that. I explored using SVD and eigen dedcomposition. First up if I remember right I believe they're paper made an error bc their formula doesn't work with SVD it only works with an eiegen decomposition.

I experimented on cora, and even using the largest 1000 (out of 2700ish) eigenvectors the resulting reconstructed matrix had large errors that caused useless embeddings. I think this was because most of the eigenvalues were the same magnitude so there were no negligible eigenvalues to discard.  AFAIK the paper didn't actually test this idea out but only mentioned it.

Also, with the current tensorflow dataset when training on GPU the pre-processing of a batch on CPU (calculating the powers for the batch rows) takes approximately the same amount of time as training on a batch on GPU, so there's not a significant amount of time added by calculating the powers in parallel to training.

Reply:
I think the training scales as `N^2` but as a ratio to GPU training CPU pre-processing will linearly increase with the number of nodes.. with the reddit dataset (200K nodes) the GPU training is still about the same as the CPU pre-processing but you're right that with larger graphs it is likely that CPU pre-processing will dominate the computation time.

Reply:
Interesting.

> I think the training scales as N^2 

Out of curiosity, is it O(N^2) or O(E) (since E is proportional to N^2 in "dense" graphs).

Reply:
I'm think the training scales as O(N^2) because it computes the dot product of every left vector with every right vector i.e. compares each node to every other node.